zerver/lib/streams.py

```diff
@@ -757,27 +757,57 @@ def can_access_stream_history_by_id(user_profile: UserProfile, stream_id: int) -
     return can_access_stream_history(user_profile, stream)
 
 
-def can_remove_subscribers_from_stream(
-    stream: Stream, user_profile: UserProfile, sub: Subscription | None
+def bulk_can_remove_subscribers_from_streams(
+    streams: list[Stream], user_profile: UserProfile
 ) -> bool:
-    if not check_basic_stream_access(user_profile, stream, sub, allow_realm_admin=True):
-        return False
-
     # Optimization for the organization administrator code path. We
     # don't explicitly grant realm admins this permission, but admins
     # implicitly have the can_administer_channel_group permission for
-    # all accessible channels.
+    # all channels. For channels the user cannot access, they can
+    # unsubscribe other users if they have permission to administer
+    # that channel.
     if user_profile.is_realm_admin:
         return True
 
-    group_allowed_to_remove_subscribers = stream.can_remove_subscribers_group
-    assert group_allowed_to_remove_subscribers is not None
-    return user_has_permission_for_group_setting(
-        group_allowed_to_remove_subscribers,
-        user_profile,
-        Stream.stream_permission_group_settings["can_remove_subscribers_group"],
+    user_recursive_group_ids = set(
+        get_recursive_membership_groups(user_profile).values_list("id", flat=True)
     )
 
+    # We check this before basic access since for the channels the user
+    # cannot access, they can unsubscribe other users if they have
+    # permission to administer that channel.
+    is_allowed_to_administer_all_streams = True
```

Comment:
I think a cleaner implementation would be `permission_failure_channels: Set[int]`. Even if we don't print it anywhere for now, it seems nicer to just do that computation; and then you can skip the `break` logic and just check if the set is nonempty just below. The loop here is very fast since it's just doing set checks in Python.

(I started this line of thinking just because the `administer_all_streams` framing is odd, and also having it mutate from True to False is a bit harder to follow as a pattern, since when first initialized it's possibly wrong)

Reply:
[Fixed](https://github.com/zulip/zulip/compare/ab1e59bfb9cc9bf0dc437de81f467709284ef93c..bedb600897f6e8efb7039117b7d3962128973f90)